app/mender.go

```diff
@@ -133,153 +136,88 @@ func NewMender(config *conf.MenderConfig, pieces MenderPieces) (*Mender, error)
-		return true
-	}
-
-	if !m.authMgr.HasKey() {
-		log.Debugf("Needs keys")
-		return true
-	}
-
-	return false
-}
-
-func (m *Mender) Bootstrap() menderError {
-	if !m.needsBootstrap() {
-		return nil
-	}
-
-	return m.doBootstrap()
-}
-
 // cache authorization code
 func (m *Mender) loadAuth() menderError {
 	if m.authToken != noAuthToken {
 		return nil
 	}
 
-	code, err := m.authMgr.AuthToken()
-	if err != nil {
-		return NewFatalError(errors.Wrap(err, "failed to cache authorization code"))
+	inChan := m.authManager.GetInMessageChan()
+	respChan := make(chan AuthManagerResponse)
+
+	// request
+	inChan <- AuthManagerRequest{
+		Action:          ActionGetAuthToken,
+		ResponseChannel: respChan,
+	}
+
+	// response
+	resp := <-respChan
+	if resp.Error != nil {
+		return NewTransientError(resp.Error)
 	}
 
-	m.authToken = code
+	m.authToken = resp.AuthToken
 	return nil
 }
 
 func (m *Mender) IsAuthorized() bool {
-	if m.authMgr.IsAuthorized() {
-		// AuthToken is present in store
-
-		if err := m.loadAuth(); err != nil {
-			return false
-		}
-		log.Info("authorization data present and valid")
+	if err := m.loadAuth(); err != nil {
+		return false
+	}
+	if m.authToken != noAuthToken {
 		return true
 	}
 	return false
 }
 
 func (m *Mender) Authorize() menderError {
-	var rsp []byte
-	var err error
-	var server *client.MenderServer
-
-	if m.authMgr.IsAuthorized() {
+	if m.IsAuthorized() {
 		log.Info("authorization data present and valid, skipping authorization attempt")
 		return m.loadAuth()
 	}
 
-	if err := m.Bootstrap(); err != nil {
-		log.Errorf("Bootstrap failed: %s", err)
-		return err
-	}
+	return m.authorize()
+}
 
-	// Cycle through servers and attempt to authorize.
-	m.authToken = noAuthToken
-	serverIterator := nextServerIterator(m)
-	if serverIterator == nil {
-		return NewFatalError(errors.New("Empty server list in mender.conf!"))
-	}
-	if server = serverIterator(); server == nil {
-		return NewFatalError(errors.New("Empty server list in mender.conf!"))
-	}
-	for {
-		rsp, err = m.authReq.Request(m.api, server.ServerURL, m.authMgr)
+func (m *Mender) authorize() menderError {
+	inChan := m.authManager.GetInMessageChan()
+	broadcastChan := m.authManager.GetBroadcastMessageChan(authManagerChannelName)
+	respChan := make(chan AuthManagerResponse)
 
-		if err == nil {
-			// SUCCESS!
-			break
-		}
-		prevHost := server.ServerURL
-		server = serverIterator()
-		if server == nil {
-			break
-		}
-		log.Warnf("Failed to authorize %q; attempting %q.",
-			prevHost, server.ServerURL)
-	}
-	if err != nil {
-		// Generate and report error.
-		errCause := errors.Cause(err)
-		if errCause == client.AuthErrorUnauthorized {
-			// make sure to remove auth token once device is rejected
-			if remErr := m.authMgr.RemoveAuthToken(); remErr != nil {
-				log.Warn("can not remove rejected authentication token")
-			}
-		}
-		return NewTransientError(errors.Wrap(err, "authorization request failed"))
+	// request
+	inChan <- AuthManagerRequest{
```

Comment:
I just realized: I think the broadcast channel should be emptied here, so that you don't risk getting the message from an old broadcast.

Reply:
done